tests/unit/core.js

```diff
@@ -183,24 +183,109 @@ exports.testNewNonNativeObject = function (test) {
   test.done();
 };
 
+exports["undefined as variable name"] = function (test) {
 
-/**
- * Test that JSHint allows `undefined` to be a function parameter.
- * It is a common pattern to protect against the case when somebody
- * overwrites undefined. It also helps with minification.
- *
- * More info: https://gist.github.com/315916
- */
-exports.testUndefinedAsParam = function (test) {
-  var code  = '(function (undefined) {}());';
-  var code1 = 'var undefined = 1;';
+  TestRun(test)
+    .test("var undefined;");
 
-  TestRun(test).test(code);
+  TestRun(test)
+    .test("let undefined;", { esnext: true });
+
+  TestRun(test)
+    .addError(1, "const 'undefined' is initialized to 'undefined'.")
```

Comment:
I wonder if we should remove this error for the `undefined` label. it puts you in a spot, because if you were doing this in es6, you would really want to use `const` and not `let`.. but then with const you get an error assigning a value and an error not assigning a value.. ? WDYT?